\cleardoublepage 

# (APPENDIX) Appendix {-}

# Exploring more about R {#more}

Chapter \@ref(about-R) provided an overview of the R language and its RStudio development environment. It also covered the different methods for installing packages and importing datasets. However, there are other features that can be run directly from the R console that are worth exploring.

## Prepare the working directory

When working on an important project, maintaining an organized structure is essential (see Section \@ref(folders)). It is advisable to structure your work into folders and subfolders containing the different components of the project (Figure \@ref(fig:organization)). Both the R project file (`.Rproj`) and related documents, such as R Markdown files (`.Rmd`), should be located within this environment.

In this context, it is not necessary to manually define the working directory, as R automatically sets it when you open the project. However, it is always useful to check the path you are using, and R itself offers the `getwd()` function that shows your working directory. \index{getwd function}


```{r, warning=FALSE}
getwd()
```


However, if you are just starting to use this software, the working path you need may not be set by default. In that case, you must tell R which directory you want to use. To do this, you can use the `setwd()` function, specifying the directory path as an argument. \index{setwd function}

```{r,warning=FALSE, echo=TRUE,error=FALSE,include=FALSE}
# setwd("C:/another_directory")
```

## In R, everything is an object

In R, both data and any other existing elements are treated as objects. The commands you have been executing throughout the book are nothing more than manipulations of objects. You may not have noticed this, as these objects are usually associated with an identifying name, commonly referred to as a variable.

In the following example, three variables called `data_01`, `data_02`, and `data_03` are created, each containing a different numeric vector. These names are assigned to the objects using the most common assignment operator in R, `<-`. However, there are other valid ways to make assignments within the language.

```{r,warning=FALSE}

data_01 <- c(10,80,20,30) # common assignment

c(5,40,15,2) -> data_02  # unusual assignment
  
data_03 = c(50,30,78,14) # valid but confusing

data_01

```

There are other ways to assign names:

- `->`: Useful for inverted reading or aesthetics.

- `=`: Valid, but can be confusing because it is also used for named arguments in functions.

## The power of R lies in its functions

Functions, also called commands, are sets of instructions identified by a name that produce a result based on the arguments provided to them. R includes a wide variety of built-in functions, many of which have already been introduced throughout the book. The following table summarizes the most frequently used basic functions. 


Table: (\#tab:functions) Examples of frequently used built-in functions in R.

| Function 	| Description 	| Example 	|
|---	|---	|---	|
| `c()` 	| Combine several elements into a vector. 	| `c(1,2,3)` #output 1,2,3 	|
| `mean()` 	| Calculates the mean (average) of a numeric vector. 	| `mean(c(2, 10, 4))` #output 5.3  	|
| `summary()` 	| Generates a statistical summary of an object. 	| `summary(c(1, 2, 3, 4, 5))` #output Min, 1st Qu.... 	|
| `sum()` 	| Calculate the sum of the values of a vector. 	| `sum(c(10, 36, 80))` #output 126 	|
| `round()` 	| Rounds a number to the specified number of digits. 	| `round(3.1416, 2)` #output 3.14 	|
| `paste()` 	| Combines texts or elements in a string. 	| `paste("I'm", "a", "biologist")` #output "I'm a biologist"  	|
| `paste0()` 	| Similar to `paste()`, but without spaces between elements. 	| `paste0("B", 1:3)` #output "B1" "B2" "B3" 	|
| `head()` 	| Displays the first 6 rows of an object (by default). 	| `head(iris,4)` # display four rows of the dataset. 	|


The Table \@ref(tab:functions) above includes only a small sample of the many [functions available](https://www.datacamp.com/doc/r/functions) in R. In addition to the functions built into the language, there are others that come included in packages that are loaded during a session. These functions are designed to facilitate specific tasks.

For example, the ggplot2 package offers a set of functions that allow you to create a wide variety of graphs (see Section \@ref(ggplot)). Similarly, the dplyr package (see Section \@ref(dplyr)) provides functions focused on efficient data manipulation.

However, it is possible to define your own functions to automate specific tasks. This has benefits, such as avoiding unnecessary code repetition, which in turn reduces the likelihood of errors. In fact, if you have noticed, a custom theme called `theme_us_classic()` has been applied to the graphics throughout the book (see Section \@ref(theme-plot)). \index{theme\_us\_classic, book}

```{r,warning=FALSE, message=FALSE, error=FALSE,eval=FALSE}

# Theme used in the book

theme_us_classic <- function(base_size  = 14, 
                     base_family  = "") {
  color = "grey50"
  line_size = 0.5
  ggplot2::theme_classic(base_size = base_size, 
    base_family = base_family) %+replace%
    theme(
    axis.text = element_text( color = color), 
    axis.line = element_line(size = line_size
    , color = color),
    axis.ticks = element_line(color = color)
    )
}
```

This customization corresponds to a function, which has a defined structure:

- **Arguments or parameters**: These are the values indicated within `function(...)`, represented by variable names such as `base_size` and `base_family`.

- **Function body**: This is the set of instructions enclosed in curly brackets `{}`. This block describes everything the function must execute and can include multiple lines of code.

Now that we understand the structure of a function, we will proceed to create one that calculates a person's Body Mass Index (BMI) and classifies it according to the criteria established by the World Health Organization ([WHO](https://iris.who.int/handle/10665/42330)).

```{r,warning=FALSE}

# Function to calculate Body Mass Index (BMI) 
# and classify according to WHO.

calculate_bmi <- function(weight, height) {
  # weight in kilograms (kg)
  # height in meters (m)
  bmi <- weight / (height ^ 2)
  
  classification <- if (bmi < 18.5) {
    "Underweight"
  } else if (bmi < 24.9) {
    "Normal weight"
  } else if (bmi < 29.9) {
    "Overweight"
  } else if (bmi < 34.9) {
    "Obesity class I"
  } else if (bmi < 39.9) {
    "Obesity class II"
  } else {
    "Obesity class III (morbid obesity)"
  }
  
  list(BMI = round(bmi, 2), 
       Classification = classification)
}

result <- calculate_bmi(weight = 70, height = 1.80)
cat("BMI:", result$BMI, "\nClassification:", result$Classification)

```

This new feature adequately fulfills its purpose, as it allows for the correct calculation of the BMI of a person weighing 70 kg and measuring 1.80 m in height.

```{block2, variable, type='rmdtip'}

__Want to know more?__

- It is recommended to use clear and descriptive names when defining your variables. These names can contain letters, numbers, and underscores, as in the case of `data_01`. In my opinion, avoid including periods (`.`) in variable names, as they can cause confusion.  

- Functions have names, and according to the [Tidyverse style guide](https://style.tidyverse.org/), they should be named using verbs (e.g., `calculate_bmi`).

```

# Markdown Tables {#table-markdown}

Section \@ref(table) explained how to generate tables using various R packages, such as [kableExtra](https://cran.r-project.org/web/packages/kableExtra/index.html), which allows you to customize your layout in both PDF and HTML formats. However, if you are producing reproducible documents with RMarkdown (see Section \@ref(reproducible)), you can also build tables directly using the extended syntax provided by Markdown (Table \@ref(tab:example)). Its basic structure and other details are shown below:

Table: (\#tab:example) Example of a table created directly from Markdown syntax.

| Description | Core     |     p-value         | 
| :---------- | :------: |     --------:       | 
| First text  | 4.2      | 0.2                 |
| Second text | 0.8      | **0.5**             | 
| *Third text*| 5        | 0.6                 |

- Vertical bars (`|`) and three (or more) dashes (`---`) can be used to create the table header.

- The first and second rows represent the column names and alignment, respectively. The other rows represent the content (data) of the table. 

- Adding a colon (`:`) to the dashes allows you to align the text in the columns to the left (`:---`), to the right (`---:`), or in the center (`:---:`).

- Make sure that each row begins and ends with a vertical bar.

```         
| Description | Core     |     p-value         | 
| :---------- | :------: |     --------:       | 
| First text  | 4.2      | 0.2                 |
| Second text | 0.8      | **0.5**             | 
| *Third text*| 5        | 0.6                 |

```

```{block2, table, type='rmdtip'}

__Want to know more?__

- The [Table Generator](https://www.tablesgenerator.com/markdown_tables) is a free online application that allows you to easily create tables in Markdown format using a visual interface.

```

# Tools for scientific illustrations {#tool}

Scientific publications do not only include graphs or tables; they often also present scientific illustrations of high visual quality. These images have the ability to communicate findings in a clear and intuitive way. However, producing them represents a challenge, as it requires researchers to have some knowledge of graphic design.

It is precisely this need to generate understandable and visually effective illustrations that has driven, in recent years, the development of several specialized tools for scientific illustration.

[BioRender](https://www.biorender.com/) is an online platform that provides users with an extensive collection of over 40,000 templates and icons specifically designed to represent biological and medical concepts. \index{BioRender} In addition, it features an intuitive drag-and-drop interface for customization according to the user's need. However, although it offers free access for basic educational purposes, a paid subscription is required to use the figures in publications [@perkel10]. 

Additionally, [PhyloPic](https://www.phylopic.org/) is a database that offers vector silhouette images of various species, each of which is associated with a taxonomic name. The platform currently has 11,209 images and they are frequently used in phylogenetic visualizations. In addition, there is the R package [rphylopic](https://rphylopic.palaeoverse.org/index.html) which allows you to add these silhouettes to R-based generated graphs and ggplot2 [@Gearty2023]. \index{rphylopic} The ggtree package [@Yu3] allows you to use PhyloPic to annotate phylogenetic trees (Figure \@ref(fig:phylopic)).


(ref:phylopic) Example of a phylogenetic tree with PhyloPic silhouettes. With the function geom="phylopic" of the ggtree package, it is possible to incorporate these silhouettes using the UID previously queried. Note: Graph based on the publication by @Yu2.

```{r phylopic, fig.width = 5.5*6/4.2, fig.asp = 0.7,dpi=300,echo=FALSE,error=FALSE, warning=FALSE, message=FALSE, fig.cap= '(ref:phylopic)', fig.align='center'}

library(ggtree)
library(tidyverse)
library(ggimage)


tr <- read.tree(text = "((a,(b,(c,d),e)),f);")

genus <- c("Pongo", "Gorilla", "Pan", "Pan", "Homo", "Nomascus")
species <- c("abelii", "gorilla", "paniscus", 
             "troglodytes", "sapiens", "leucogenys")
subspecies <- c("","gorilla", "","","","")

name <- c("Pongo_abelii", "Gorilla_gorilla_gorilla", "Pan_paniscus",
            "Pan_troglodytes", "Homo_sapiens", "Nomascus_leucogenys") 

p <- data.frame(label = tr$tip.label, genus = genus,
                species = species, name = name, 
                subspecies = subspecies)

names <- paste0("((Pongo_abelii,(Gorilla_gorilla_gorilla,(Pan_paniscus,",
          "Pan_troglodytes)Pan,Homo_sapiens)Homininae)Hominidae,",
          "Nomascus_leucogenys)Hominoidea;")

tree <- read.tree(text=names)

p2 <- ggimage::phylopic_uid(tree$tip.label)

data <- full_join(p, p2, by = 'name')

ggtree(tr,color='firebrick', size = 1) %<+% data  +
geom_tiplab(aes(label=paste0(
'italic(', genus, ')~~italic(', species, ')~~italic(', subspecies, ')')), parse=T) + 
geom_tiplab(aes(image=uid), geom="phylopic", offset=2) +
xlim(NA, 7)


```


The Generic Diagramming Platform ([GDP](https://biogdp.com/)) is a recently developed tool that offers an extensive collection of high quality biomedical images. \index{GDP} It currently has more than 8000 graphics classified in various categories and, thanks to its interactive interface, allows the creation of professional illustrations suitable for publication [@Jiang2025].

However, if your goal is to render three-dimensional visualizations (e.g. molecular surfaces), [Blender](https://www.blender.org/) is the best alternative [@Zini2010;@Spalvieri2022]. \index{Blender} It is a free and open source software that allows you to process visualization including: assembly, animation, simulation, rendering, and more [@illustrae]. When you first start using it, you may find the tool intimidating due to the number of panels and options available. However, thanks to its large and active community, it is easy to find free tutorials and workshops that make it easy to learn for those new to Blender [@Parlikar10].

Adobe Illustrator is a commercial vector design tool that allows you to zoom in and out of visualizations without loss of quality. It is a popular choice among scientists for producing scientific illustrations, posters, aligned panels and graphical abstracts suitable for publication [@Jambor2024]. However, if you need to work with images in a more advanced way, including artificial intelligence (AI)-driven features, Adobe Creative Cloud represents an alternative, although it requires a prior subscription [@illustrae].

There are also free and open source tools, such as [Inskcape](https://inkscape.org/) [@Jambor2024]. \index{Inskcape} This graphic editor offers a complete set of functions for the creation and editing of vector images, being on a par with Adobe Illustrator, therefore it has gained popularity among researchers (Figure \@ref(fig:tool)).

(ref:tool) Inskcape interface. Its easy-to-use interface, variety of tools, its compatibility with numerous file formats and free of charge, has allowed it to gain popularity among scientists.

```{r tool, out.width='100%',dpi=300,echo=FALSE,warning=FALSE, message=FALSE, fig.cap= '(ref:tool)', fig.align='center'}

knitr::include_graphics(rep("images/inskcape.png"))
```


Adobe Illustrator and Inkscape can play a key role in the post-processing phase of data visualizations. This process typically follows a workflow that encompasses the following stages:

- First, the user generates his plots with visualization software, such as R using the ggplot2 package, and exports them in vector formats such as PDF or SVG (see Chapter \@ref(export-plot)).

- The exported files are then opened in editing programs such as Inkscape, where additional modifications are made, such as relocating labels, adjusting color, or adding annotations, all without modifying the original visual content. Inkscape, in particular, has an extension called [Scientific Inkscape](https://github.com/burghoff/Scientific-Inkscape?tab=readme-ov-file), which facilitates efficient importing, editing and scaling of scientific figures. \index{Inskcape!Scientific Inkscape}

- Finally, the final figure is exported in formats compatible for publication (e.g. PDF).


```{block2, blender, type='rmdtip'}

__Want to know more?__

- The film Flow (2024), winner of the Oscar for Best Animated Feature Film in 2025, was rendered entirely in [Blender](https://www.blender.org/) with a running time of approximately 84 minutes.

- The work published by @McNair2023 offers a detailed guide on how to generate plant silhouettes and upload them to the PhyloPic platform. It also incorporates explanatory videos that present various techniques for making them.

```

# About the Author {#author}

```{r, out.width='48%',dpi=300,echo=FALSE,warning=FALSE, message=FALSE, fig.align='center'}

knitr::include_graphics(rep("images/WVC.jpg"))
```

**William Guillermo Vilchez Cruz** holds a bachelor's degree in Biological Sciences from the [Pedro Ruiz Gallo National University](https://www.unprg.edu.pe/univ/portal/index.php), located in Lambayeque, Peru. He is currently pursuing a master's degree in Cellular and Molecular Biology at the [University of SÃ£o Paulo](https://www5.usp.br/) (USP) in Brazil. Prior to entering USP, he worked as a laboratory assistant in the Molecular Biology course at his alma mater, an experience that gave him the opportunity to collaborate on various student-led research projects.

Based on these academic experiences, and motivated by the challenges faced by many researchers and students in representing complex data clearly and effectively, this book emerges as a proposal aimed at improving data visualization skills through the use of the ggplot2 package of the R language.

The author is open to receiving suggestions and constructive comments that contribute to improving the content and focus of this work.

He can be contacted via email: wvilchezc@unprg.edu.pe or through his social media accounts.

[Github](https://github.com/WVC20-bio)/ [Twitter](https://x.com/WilliamVilche17)
